### PR TITLE
#8716: introduce constants for renamed strings

### LINKED ISSUE #8716: Introduce `Kinds` constant
On several different types, we have a `kind` property that always one of e.g. ["extensionPoint", "component", "recipe", "service", "reader"], most of which are legacy terms.

Let's introduce a `KINDS` constant that we can use in type definitions and logic that map to non-legacy terms, with the goal of eliminating "magic strings" and improving the readability of code with legacy terminology that we can't change in the short-term.

```
Kinds = {
    STARTER_BRICK: "extensionPoint",
    MOD: "recipe",
    INTEGRATION: "service",
    BRICK: "component",
    READER: "reader"
}
```

[registryTypes.ts](https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/types/registryTypes.ts#L27-L27) is likely a good candidate for a file to put this in.

**Kinds of `kind`**
- The `kind` property that originates from yaml definitions (e.g. https://github.com/pixiebrix/pixiebrix-extension/tree/main/schemas) and can be one of `["extensionPoint", "component", "recipe", "service", "reader"]`
- The `kind` property that originates from the back-end database representation of `Package.kind` (e.g. https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/types/contract.ts#L140-L140) which can be one of `"Blueprint" | "Service" | "Block" | "Reader" | "Foundation"`
    - Less common in terms of "magic string" occurrences, but we might benefit from also creating a constant for this e.g. `const PackageKind = { MOD: "Blueprint", INTEGRATION: "Service", ...`
- Mapped versions of `kind` to user-facing terminology (see prior art) usually seen in Workshop contexts, can be one of `"Brick" | "Mod" | "Integration" | "Starter"`

**Other magic string elimination opportunities**
- The related PageState `Namespace` type (see https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/bricks/effects/pageState.ts#L30-L30) which can be one of `"blueprint" | "extension" | "shared"`
    - Introduce a `Namespace` constant e.g. `const Namespace = {MOD: "blueprint", MOD_COMPONENT: "extension", PUBLIC: "shared}` 

**Prior Art**
- [mapKindToUIValue](https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/extensionConsole/pages/workshop/workshopUtils.ts#L33-L33)
- [`KIND_SCHEMAS` map](https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/validators/schemaValidator.ts#L71-L71)
- [`Kind` type](https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/types/registryTypes.ts#L37-L42)
- [Similar work for telemetry events](https://github.com/pixiebrix/pixiebrix-extension/blob/main/src/telemetry/events.ts#L23-L23)

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Closes #8716
- Follows pattern of `Events` (using object instead of enum)
- Introduces `DefinitionKinds`, `BrickTypes`, `StateNamespaces`
- Drops Kind type from PackageRegistry which contains some non-sense values

## Discussion/Reviewer Notes
- Introduces all 3 in the same PR because otherwise occurrences of "reader" and "blueprint" are ambiguous between higher-level kinds. So it's hard to lexically search for occurrences of the raw string
- Uses "kind" because it corresponds to "Definition.kind". Uses "type" because brick that the term existing code uses, e.g., TypedBrickPair

For more information on our expectations for the PR process, see the
[code review principles doc](https://www.notion.so/pixiebrix/Code-Review-Principles-1ce7276b82a84d2a995d55ad85e1310d?pvs=4)

### CODE REVIEW

File: src/types/registryTypes.ts [hunk 1]
```diff
@@ -32,14 +33,22 @@ export type RegistryId = string & {
...
+export const DefinitionKinds = {
+  STARTER_BRICK: "extensionPoint",
+  MOD: "recipe",
+  INTEGRATION: "service",
+  BRICK: "component",
+  READER: "reader",
+} as const;
+
+/**
+ * The type for of definition in the external package registry.
```
Comment (mnholtz) [inline src/types/registryTypes.ts:49]:
```suggestion
 * The type for definitions in the external package registry.
```

File: src/types/registryTypes.ts [hunk 2]
```diff
@@ -32,14 +33,22 @@ export type RegistryId = string & {
 export const INNER_SCOPE = "@internal";
 
 /**
- * The kind of definition in the external registry
+ * Constants for the different kinds of top-level definitions in the registry.
+ *
+ * See https://github.com/pixiebrix/pixiebrix-extension/tree/main/schemas
+ */
+export const DefinitionKinds = {
```
Comment (mnholtz) [inline src/types/registryTypes.ts:40]: Love this name choice.

File: src/registry/packageRegistry.ts [hunk 3]
```diff
@@ -35,21 +36,10 @@ type Version = {
...
-  | "effect"
-  | "component"
-  | "extensionPoint"
-  | "recipe";
-
 export type PackageVersion = {
   id: string;
   version: Version;
-  kind: Kind;
+  kind: DefinitionKind;
```
Comment (mnholtz) [inline src/registry/packageRegistry.ts:42]: What is going on here with the implicit omission of `"block"`, `"foundation"`, `"blueprint"` and `"effect"`? Were those never valid kinds to begin with here?

File: src/extensionConsole/pages/brickEditor/useSubmitBrick.ts [hunk 4]
```diff
@@ -126,9 +126,9 @@ function useSubmitBrick({ create = false }: SubmitOptions): SubmitCallbacks {
         // We attach the handler below, and don't want it to block the save
         void (async () => {
           try {
-            if (kind === "recipe" && reinstallBlueprint) {
-              // TypeScript doesn't have enough information to kind === "recipe" distinguishes ModDefinition from
-              // Definition
+            if (kind === DefinitionKinds.MOD && reinstallBlueprint) {
+              // TypeScript doesn't have enough information to kind === PackageKinds.MOD distinguishes ModDefinition
```
Comment (mnholtz) [inline src/extensionConsole/pages/brickEditor/useSubmitBrick.ts:130]:
```suggestion
              // TypeScript doesn't have enough information to infer that kind === PackageKinds.MOD distinguishes ModDefinition
```

File: src/bricks/registry.ts [hunk 5]
```diff
@@ -39,7 +39,7 @@ export type TypedBrickMap = Map<RegistryId, TypedBrickPair>;
  */
 class BrickRegistry extends MemoryRegistry<RegistryId, Brick> {
   constructor() {
-    super(["block", "component", "effect", "reader"], null);
+    super(["component", "reader"], null);
```
Comment (mnholtz) [inline src/bricks/registry.ts:42]: Why not replace these strings also?

File: src/registry/packageRegistry.ts (hunk 3 above)
Comment (mnholtz) [reply src/registry/packageRegistry.ts:42]: > I want to get this PR merged sooner than later though to see if there's something funky going on in the wild

For context, are you suspecting this is tied to a bug?

**Review by mnholtz — APPROVED** [review]
amazing

File: src/bricks/registry.ts [hunk 6]
```diff
@@ -39,7 +39,8 @@ export type TypedBrickMap = Map<RegistryId, TypedBrickPair>;
  */
 class BrickRegistry extends MemoryRegistry<RegistryId, Brick> {
   constructor() {
-    super(["block", "component", "effect", "reader"], null);
+    // A reader is a special kind of brick. So we keep them in the same in-memory registry.
```
Comment (mnholtz) [inline src/bricks/registry.ts:42]: This might be a good comment for reader prop on DefinitionKinds

File: src/registry/packageRegistry.ts (hunk 3 above)
Comment (twschiller) [reply src/registry/packageRegistry.ts:42]: > For context, are you suspecting this is tied to a bug

No, I just want to make sure I'm not missing something and those values could actually be passed to the registry

**Pull request merged**